pkg/cmd/server/kubernetes/master_config.go

```diff
@@ -106,7 +93,95 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 	// TODO: this should be done in config validation (along with the above) so we can provide
 	// proper errors
 	if err := cmdflags.Resolve(options.KubernetesMasterConfig.APIServerArguments, server.AddFlags); len(err) > 0 {
-		return nil, kerrors.NewAggregate(err)
+		return nil, nil, kerrors.NewAggregate(err)
+	}
+
+	resourceEncodingConfig := genericapiserver.NewDefaultResourceEncodingConfig()
+	resourceEncodingConfig.SetVersionEncoding(
+		kapi.GroupName,
+		unversioned.GroupVersion{Group: kapi.GroupName, Version: options.EtcdStorageConfig.KubernetesStorageVersion},
+		kapi.SchemeGroupVersion,
+	)
+
+	resourceEncodingConfig.SetVersionEncoding(
+		extensions.GroupName,
+		unversioned.GroupVersion{Group: extensions.GroupName, Version: "v1beta1"},
+		extensions.SchemeGroupVersion,
+	)
+
+	resourceEncodingConfig.SetVersionEncoding(
+		batch.GroupName,
+		unversioned.GroupVersion{Group: batch.GroupName, Version: "v1"},
+		batch.SchemeGroupVersion,
+	)
+
+	resourceEncodingConfig.SetVersionEncoding(
+		autoscaling.GroupName,
+		unversioned.GroupVersion{Group: autoscaling.GroupName, Version: "v1"},
+		autoscaling.SchemeGroupVersion,
+	)
+
+	storageGroupsToEncodingVersion, err := server.StorageGroupsToEncodingVersion()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	etcdConfig := storagebackend.Config{
+		Type:       server.StorageConfig.Type,
+		Prefix:     options.EtcdStorageConfig.KubernetesStoragePrefix,
+		ServerList: options.EtcdClientInfo.URLs,
+		KeyFile:    options.EtcdClientInfo.ClientCert.KeyFile,
+		CertFile:   options.EtcdClientInfo.ClientCert.CertFile,
+		CAFile:     options.EtcdClientInfo.CA,
+		DeserializationCacheSize: genericapiserveroptions.DefaultDeserializationCacheSize,
+		Quorum: server.StorageConfig.Quorum,
+	}
+
+	storageFactory, err := genericapiserver.BuildDefaultStorageFactory(
+		etcdConfig,
+		server.DefaultStorageMediaType,
+		kapi.Codecs,
+		genericapiserver.NewDefaultResourceEncodingConfig(),
+		storageGroupsToEncodingVersion,
+		// FIXME: this GroupVersionResource override should be configurable
+		[]unversioned.GroupVersionResource{batch.Resource("scheduledjobs").WithVersion("v2alpha1")},
+		master.DefaultAPIResourceConfigSource(), server.RuntimeConfig,
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	/*storageFactory := genericapiserver.NewDefaultStorageFactory(
+		etcdConfig,
+		server.DefaultStorageMediaType,
+		kapi.Codecs,
+		resourceEncodingConfig,
+		master.DefaultAPIResourceConfigSource(),
+	)*/
+	// the order here is important, it defines which version will be used for storage
+	storageFactory.AddCohabitatingResources(extensions.Resource("jobs"), batch.Resource("jobs"))
```

Comment:
This might be problematic soon, since I'm planning to remove Jobs from extensions in k8s 1.5, see https://github.com/kubernetes/kubernetes/issues/32763